src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+//! An in-memory data store for Merkle-lized data
+//!
+//! This is a in memory data store for Merkle trees, this store allows all the nodes of a tree
+//! (leaves or internal) to live as long as necessary and without duplication, this allows the
+//! implementation of efficient persistent data structures
+use super::{
+    BTreeMap, Direction, MerkleError, MerklePath, MerkleTree, NodeIndex, Rpo256, RpoDigest,
+    SimpleSmt, Vec, Word,
+};
+
+// FIX: This is a temporary work around for the lack of Ord in BaseElement
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct WordWithOrd {
+    word: Word,
+}
+
+impl PartialOrd for WordWithOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.partial_cmp(other_ints)
+    }
+}
+
+impl Ord for WordWithOrd {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.cmp(other_ints)
+    }
+}
```

Comment:
> The reason Ord isn't implemented for Word is because their inner element isn't in canonical form; hence, ordering isn't sound.

humm, I didn't think about it too much, but for the usage we have in here it should be correct, values which are equal hash to the same value. But I can do the change if necessary.

Reply:
Ditto; for a quickfix we can postpone the canonical word issue. However, it is important we use `as_int` instead of `inner` because otherwise the ordering will be broken

Reply:
Moved to https://github.com/0xPolygonMiden/crypto/pull/93#discussion_r1134913337